### PR TITLE
Improvement in Gadget Generation wizard

### PR DESCRIPTION
**Pull request by Subangani**
Grouping of provider's config file and add the functionality of editing the gadgets and giving warning message before editing.

### CODE REVIEW

File: apps/portal/controllers/apis/createGadget.jag [hunk 1]
```diff
@@ -104,6 +110,13 @@
         }
         print(customProviderPartial);
         return;
+    }
+    else if (action==='getEditConfig'){
```
Comment (sinthuja) [inline apps/portal/controllers/apis/createGadget.jag]: There should be a space before and after ====/

File: apps/portal/controllers/apis/createGadget.jag [hunk 2]
```diff
@@ -104,6 +110,13 @@
...
+        providerApiLocation = GADGET_LOCATION + store +'/'+type +'/'+ id+ EDIT_CONFIG_FILE;
```
Comment (sinthuja) [inline apps/portal/controllers/apis/createGadget.jag]: Do code formatting, and space.

File: apps/portal/controllers/apis/createGadget.jag [hunk 3]
```diff
@@ -104,6 +110,13 @@
...
+        providerApi = require(providerApiLocation);
+        print(providerApi);
+        return;
+
```
Comment (sinthuja) [inline apps/portal/controllers/apis/createGadget.jag]: Remove unnecessary lines.

File: apps/portal/controllers/apis/createGadget.jag [hunk 4]
```diff
@@ -223,7 +236,7 @@
             gadgetID = chartConf[GADGET_NAME].replace(/ /g,"_");
 
             var gadgetStoreFile = new File(GADGETS_LOCATION + gadgetID);
-            if(gadgetStoreFile.isExists()){
+            if(gadgetStoreFile.isExists() && editable!=='true' ){
```
Comment (sinthuja) [inline apps/portal/controllers/apis/createGadget.jag]: Do code formatting, and space.

File: apps/portal/controllers/apis/createGadget.jag [hunk 5]
```diff
@@ -362,6 +375,16 @@
         if(providerApi.getMode().toLowerCase() === 'push'){
             isPushMode = true;
         }
+        var gadgetStoreFile = new File(GADGETS_LOCATION + gadgetID);
+
```
Comment (sinthuja) [inline apps/portal/controllers/apis/createGadget.jag]: Cleanup code by removing unnecessary lines.

File: apps/portal/js/wizard.js [hunk 6]
```diff
@@ -288,40 +298,69 @@ $(document).on('click', '.complete', function () {
...
+                $("#provider-list").html(providerHbs(data));
+            },
+            error: function (xhr, message, errorObj) {
+                //When 401 Unauthorized occurs user session has been log out
+                if (xhr.status == 401) {
+                    //reload() will redirect request to login page with set current page to redirect back page
+                    location.reload();
+                }
+                var source = $("#wizard-error-hbs").html();
+                ;
```
Comment (sinthuja) [inline apps/portal/js/wizard.js]: Remove unnecessary semi colon.

File: apps/portal/js/wizard.js [hunk 7]
```diff
@@ -432,10 +536,43 @@ function getChartConfig(providerConfig) {
...
+                    configGroup = {
+                        "grouping": {
+                            "elements": chartConfig
+                        }
+                    };
+                    if (editable) {
+                        var gadgetCong = getGadgetConf();
+                        var group = addValueWithoutGrouping(configGroup, gadgetCong['chart-conf']);
+                        registerAdvancedProviderUI(configGroup);
+                        console.log(group);
```
Comment (sinthuja) [inline apps/portal/js/wizard.js]: Remove console.log.

File: apps/portal/theme/templates/create-gadget.jag [hunk 8]
```diff
@@ -28,14 +28,18 @@
     <link rel="stylesheet" href="<%=urlPrefix%>css/wizard.css">
     <link rel="stylesheet" href="<%=urlPrefix%>css/d3.geomap.css">
     <link rel="stylesheet" href="<%=urlPrefix%>css/theme.css">
-    
-    <% include('/controllers/includes/theme-portal-styles.jag'); %>  
-    
+
+    <% include('/controllers/includes/theme-portal-styles.jag');
```
Comment (sinthuja) [inline apps/portal/theme/templates/create-gadget.jag]: Add a space to separate variable declarations.

File: apps/portal/js/assets.js [hunk 9]
```diff
@@ -319,6 +334,44 @@ $(function () {
...
+                assetElement.html(editGadgetUsageConfirmHbs(data));
+            }
+        });
+
+        portal.on('click', '.ds-assets .ds-asset-edit-confirm', function (e) {
+            e.preventDefault();
+            var assetElement = $(this).closest('.ds-asset');
+            var id = assetElement.data('id');
+            var store = $('#selectStore').selectpicker('val');
+            if (typeof(store) !== "string") {
```
Comment (udarakr) [inline apps/portal/js/assets.js]: our practice is to use (typeof store !== 'string')

File: apps/portal/css/asset.css [hunk 10]
```diff
@@ -66,6 +66,10 @@ body {
     text-align: right;
     white-space: nowrap;
 }
+.glyphicon-pencil{
```
Comment (udarakr) [inline apps/portal/css/asset.css]: add newline before line 69

File: apps/portal/js/wizard.js [hunk 11]
```diff
@@ -288,40 +298,68 @@ $(document).on('click', '.complete', function () {
...
+        var providerHbs = Handlebars.compile($('#datasource-providers-hbs').html());
+        $("#provider-list").html(providerHbs(data));
+    } else {
+        $.ajax({
+            url: ues.utils.relativePrefix() + CREATE_GADGET_API + '?action=getProviders',
+            method: "GET",
+            async: false,
+            contentType: "application/json",
+            success: function (data) {
+                if (data.length == 0) {
```
Comment (udarakr) [inline apps/portal/js/wizard.js]: use ===, data.length === 0

File: apps/portal/js/wizard.js [hunk 12]
```diff
@@ -288,40 +298,68 @@ $(document).on('click', '.complete', function () {
...
+                    $("#rootwizard").empty();
+                    $("#rootwizard").append(template());
+                    return;
+                }
+                var providerHbs = Handlebars.compile($('#datasource-providers-hbs').html());
+                $("#provider-list").html(providerHbs(data));
+            },
+            error: function (xhr, message, errorObj) {
+                //When 401 Unauthorized occurs user session has been log out
+                if (xhr.status == 401) {
```
Comment (udarakr) [inline apps/portal/js/wizard.js]: use ===

File: apps/portal/controllers/apis/createGadget.jag (hunk 1 above)
Comment (Subangani) [inline apps/portal/controllers/apis/createGadget.jag]: Corrected in the commits named 'Corrections'

File: apps/portal/controllers/apis/createGadget.jag (hunk 2 above)
Comment (Subangani) [inline apps/portal/controllers/apis/createGadget.jag]: Corrected in the commits named 'corrections'

File: apps/portal/controllers/apis/createGadget.jag (hunk 3 above)
Comment (Subangani) [inline apps/portal/controllers/apis/createGadget.jag]: Corrected in the commits named 'corrections'

File: apps/portal/controllers/apis/createGadget.jag (hunk 4 above)
Comment (Subangani) [inline apps/portal/controllers/apis/createGadget.jag]: Corrected in the commits named 'corrections'

File: apps/portal/js/wizard.js (hunk 6 above)
Comment (Subangani) [inline apps/portal/js/wizard.js]: Corrected in the commits named 'corrections'

File: apps/portal/controllers/apis/createGadget.jag (hunk 5 above)
Comment (Subangani) [inline apps/portal/controllers/apis/createGadget.jag]: Corrected in the commits named 'corrections'

File: apps/portal/js/wizard.js (hunk 7 above)
Comment (Subangani) [inline apps/portal/js/wizard.js]: Corrected in the commits named 'corrections'

File: apps/portal/theme/templates/create-gadget.jag (hunk 8 above)
Comment (Subangani) [inline apps/portal/theme/templates/create-gadget.jag]: Corrected in the commits named 'corrections'

**Comment by TanyaM** [comment]
We have discontinued development of jaggery based dashboard. Hence closing the PR

**Pull request closed without merging**